File: webdash_assets/app/pages/Correlation.js
// Correlations — how many INDEPENDENT bets the race really is (the "strategy ETF"
// question), through the lens that matters for an asymmetric-payoff book:
// SIMULTANEOUS LOSSES. A favorite pays ~+0.11 to win but −1.00 to lose, so what ruins
// the book is a cluster of losses in the same window. Primary metric = φ on the per-
// window loss indicators (do two runners lose together?); full-P&L ρ is a secondary
// toggle. Redundant pairs (|φ|≥0.7) = lose together → don't double-arm.
import { html, useState } from '../preact.js';
import { usePoll } from '../api.js';
import { glyphOf, nameOf } from '../format.js';
import { Kpi, Panel, Empty, Note } from '../components/ui.js';
import { DataTable } from '../components/DataTable.js';

const rho2 = x => (x >= 0 ? "" : "-") + Math.abs(x).toFixed(2).replace(/^0/, "");

// diverging heat over the dark theme: positive (lose/co-move together) → vermilion
// (redundant, bad), negative (one wins when the other loses = a hedge) → verdigris
// (good), ~0 → faint (independent, the goal).
function cellBg(v){
  if(v == null) return "transparent";
  const a = Math.min(0.82, Math.abs(v)).toFixed(3);
  return v >= 0 ? `rgba(224,87,58,${a})` : `rgba(92,186,141,${a})`;
}

export function Correlation(){
  const data = usePoll("/correlation-data").data;
  const [view, setView] = useState("loss");   // "loss" (primary) | "pnl" | "struct"
  if(!data) return html`<div class="empty">Computing correlations…</div>`;
  const { names, matrix, losses, redundant, n_strats, n_groups, avg_abs_loss,
          n_pairs_loss, avg_abs_rho, n_pairs, struct_groups, avg_abs_struct, n_pairs_struct,
          min_overlap, min_losses, redundant_rho } = data;

  if(!n_strats) return html`<${Empty}>No strategy in the race.</${Empty}>`;
  const lossView = view === "loss", structView = view === "struct";
  const val = c => structView ? c.struct : lossView ? c.loss : c.rho;

  const idx = names.map((_, i) => i + 1);   // 1..N column heads (compact) ↔ numbered rows
  const redCols = [
    {key:"a", label:"Strategy A", cell:r=>html`<span class="cell-name"><span class="g">${glyphOf(r.a)}</span><span class="lead">${nameOf(r.a)}</span></span>`},
    {key:"b", label:"Strategy B", cell:r=>html`<span class="cell-name"><span class="g">${glyphOf(r.b)}</span><span class="lead">${nameOf(r.b)}</span></span>`},
    {key:"loss", label:"loss φ", sortable:true, get:r=>Math.abs(r.loss),
      cell:r=>html`<span class="num neg">${rho2(r.loss)}</span>`},
    {key:"both", label:"Simultaneous losses", sortable:true,
      cell:r=>html`<span class="num">${r.both}</span><span class="mut"> / ${r.la}·${r.lb}</span>`},
    {key:"n", label:"Windows", sortable:true, cell:r=>html`<span class="num mut">${r.n}</span>`},
  ];

  const toggle = html`<span class="filters" style="display:inline-flex;gap:6px">
    <button class=${"btn sm" + (lossView ? " primary" : "")} onClick=${()=>setView("loss")}>Simultaneous losses</button>
    <button class=${"btn sm" + (structView ? " primary" : "")} onClick=${()=>setView("struct")}>Structural (spot)</button>
    <button class=${"btn sm" + (view==="pnl" ? " primary" : "")} onClick=${()=>setView("pnl")}>Full P&L</button></span>`;

  return html`
    <div class="page-head">
      <div><h1><span class="glyph">▦</span>Correlations</h1>
        <p class="sub">A basket of strategies can <i>look</i> like several bets while loading
          a single risk factor. And what ruins an asymmetric-payoff book (favorite: +0.11 won /
          −1.00 lost) is <b>losses that cluster</b>: a burst of flips in the same window.
          So we measure, live, how much the runners <b>lose together</b> — the prerequisite for an
          "ETF" that would arm the top X without stacking bets that crater at the same time. The φ of realized
          losses fills in slowly (favorites ~90%); the <b>Structural</b> tab gives the answer
          <i>right away</i> via the spot co-movement of the underlyings.</p></div>
    </div>

    <div class="kpis">
      <${Kpi} k="Strategies" v=${n_strats} sub="in the race" />
      <${Kpi} accent="gold" k="Risk factors (spot)" v=${struct_groups==null ? "—" : struct_groups}
        sub=${struct_groups==null ? "spot correlation unavailable"
          : `truly independent bets — same coin or correlated spot (≥${redundant_rho}) merged · available right away`} />
      <${Kpi} k="Independent groups (losses)" v=${n_groups}
        sub=${`merging pairs that lose together (|φ|≥${redundant_rho}) — fills in slowly`} />
      <${Kpi} accent="merc" k="Loss correlation" v=${avg_abs_loss==null ? "—" : rho2(avg_abs_loss)}
        sub=${avg_abs_loss==null ? "not enough common losses yet" : `avg |φ| · ${n_pairs_loss} pairs measured`} />
      <${Kpi} k="Risk duplicates" v=${redundant.length}
        sub=${`pairs that lose together (|φ|≥${redundant_rho})`} tone=${redundant.length ? "neg" : "mut"} />
    </div>

    <${Panel} title=${structView ? "Matrix — spot co-movement of the underlyings (structural)"
        : lossView ? "Matrix — simultaneous-loss correlation (φ)" : "Matrix — full-P&L correlation (ρ)"}
      aside=${toggle}>
      <div class="heat-wrap">
        <table class="heat">
          <thead><tr><th class="rh"></th>${idx.map(i => html`<th title=${nameOf(names[i-1])}>${i}</th>`)}</tr></thead>
          <tbody>
            ${names.map((n, i) => html`<tr>
              <th class="rh" title=${nameOf(n)}><span class="g">${glyphOf(n)}</span>${i+1} · ${nameOf(n)}</th>
              ${matrix[i].map((c, j) => {
                const diag = i === j;
                const v = val(c);
                const bg = diag ? "var(--surface-3)" : cellBg(v);
                const txt = diag ? html`<span class="g">${glyphOf(n)}</span>`
                  : v == null ? html`<span class="faint">·</span>` : rho2(v);
                const ttl = diag
                  ? `${nameOf(n)} — ${c.n} settled windows, ${losses[n]} lost`
                  : structView
                    ? (v == null
                        ? `${nameOf(n)} × ${nameOf(names[j])} — spot co-movement unavailable (one of the coins has no data)`
                        : v >= 0.999
                          ? `${nameOf(n)} × ${nameOf(names[j])} — SAME underlying: structurally a single bet (simultaneous flips guaranteed)`
                          : `${nameOf(n)} × ${nameOf(names[j])} — spot correlation=${v.toFixed(2)} (co-movement of the underlyings = driver of simultaneous flips)`)
                    : lossView
                    ? (v == null
                        ? `${nameOf(n)} × ${nameOf(names[j])} — too few common losses for a reliable φ `
                          + `(they lose ${c.la}× and ${c.lb}× over ${c.n} windows, ${c.both} simultaneous; need ≥${min_losses} each)`
                        : `${nameOf(n)} × ${nameOf(names[j])} — loss φ=${v.toFixed(2)} · ${c.both} simultaneous losses `
                          + `(out of ${c.la}/${c.lb}, ${c.n} windows${c.lift!=null ? `, ×${c.lift.toFixed(1)} vs chance` : ""})`)
                    : (v == null
                        ? `${nameOf(n)} × ${nameOf(names[j])} — only ${c.n} common windows (< ${min_overlap})`
                        : `${nameOf(n)} × ${nameOf(names[j])} — ρ=${v.toFixed(2)} over ${c.n} windows`);
                return html`<td style=${"background:"+bg} title=${ttl}
                  class=${(!diag && v!=null && Math.abs(v)>=redundant_rho) ? "red" : ""}>${txt}</td>`;
              })}
            </tr>`)}
          </tbody>
        </table>
      </div>
      <${Note}>${structView
        ? html`Each cell = correlation of the <b>spot returns</b> of the two <i>underlyings</i> (recent 15m grid,
            ≈66 windows, Binance source). Why: a favorite flips when the underlying reverses within the window,
            so <b>correlated coins flip — and lose — together</b>. It's the <b>driver</b> of clustered losses,
            measurable <b>right away</b> whereas the realized-loss φ takes months to fill in (favorites ~90%).
            Same coin = <b>1.00</b> (a single bet by construction — the prudent default on the risk side). It's an <i>honest
            upper bound</i> on the loss φ (spot co-movement, not the loss itself), not a substitute once the real φ has
            filled in. The reading that matters for <b>weighting</b>: how many <i>truly</i> independent bets (the
            "Risk factors" KPI) — crypto-beta often melts 48 runners into a handful.`
        : lossView
        ? html`Each cell = φ correlation of the two runners' <b>losses</b>, window by window (loss = negative net
            P&L on the window; aligned by opening timestamp, so same <i>frame</i> = same
            grid → BTC·15m and ETH·15m are comparable). On hover: each one's losses, <b>simultaneous</b> losses,
            and the ×N factor vs chance. Favorites win ~90% → losses are rare: below ${min_overlap}
            common windows or ${min_losses} losses on each side, the cell stays empty (a φ over 1-2 losses
            is just noise). It's the most data-hungry lens — it will fill in slowly;
            meanwhile, the <b>Structural</b> tab gives the spot co-movement, available immediately.`
        : html`Each cell = Pearson correlation of the two runners' <b>full P&L</b> (ups AND downs),
            window by window. Useful to see the overall co-movement, but for the ETF it's the correlation of
            <i>losses</i> that decides (the payoff is asymmetric). Below ${min_overlap} common windows, empty.`}
      </${Note}>
    </${Panel}>

    <${Panel} title=${`Risk duplicates — lose together, keep only one (|φ|≥${redundant_rho})`} flush=${true}>
      ${redundant.length
        ? html`<${DataTable} cols=${redCols} rows=${redundant} sort=${{key:"loss",dir:-1}} empty="—" />`
        : html`<div class="panel-body"><${Note} tone="ok">No pair that systematically loses together
            so far: either the losses don't cluster, or there aren't yet enough common losing windows
            to measure it (favorites ~90% win rate → losses accumulate slowly).</${Note}></div>`}
    </${Panel}>`;
}

File: pmlab/correlation.py
"""Cross-strategy correlation — how many INDEPENDENT bets the lineup really is, with
SIMULTANEOUS LOSSES as the primary lens.

The "strategy ETF" question: a row of runners can look like many bets while loading
on ONE risk factor. And what actually ruins such a book is its asymmetric payoff — a
favorite pays ~+0.11 to win but −1.00 to lose, so a CLUSTER of losses in the same window
hurts far more than the wins help. Full-P&L Pearson treats upside and downside the same
and hides that. So the lens that drives the ETF decision here is the **simultaneous-loss
correlation**: do two runners tend to lose in the SAME window?

Method — align by RESOLVED WINDOW, then two lenses over the overlap:
  • Each runner's realized P&L is bucketed per window by window-start ts
    (journal.window_pnl). The ts is coin-agnostic, so same-frame coins share the grid
    (btc-15m & eth-15m align on every 15m boundary; a 15m ts is also a 5m ts → 5m/15m
    overlap on 1/3 of windows). A window is a LOSS for a runner iff its net P&L there < 0.
  • LOSS (primary): phi = Pearson of the two 0/1 loss-indicators over the windows BOTH
    bet — the simultaneous-loss correlation. Reported with the raw co-loss counts (how
    often each lost, how often BOTH lost) + a lift vs independence, so the coefficient is
    never read in a vacuum. Favorites win ~90%, so losses are RARE: phi needs both enough
    overlap AND enough losses each side (MIN_LOSSES) or it's noise → None.
  • P&L (secondary): plain Pearson of the per-window P&L — the full co-movement, kept as a
    toggle in the UI.
  • STRUCTURAL (the leading indicator): the realized loss-phi fills SLOWLY — favorites win
    ~90 %, so a co-loss pair needs months to clear MIN_LOSSES. But the simultaneous-flip
    risk is DRIVEN by the underlyings' spot co-movement: when the market dumps in a window,
    correlated coins reverse together and their favorites flip together. That co-movement is
    measurable NOW from spot, with no need for losses to accumulate. So we also lift a
    coin×coin spot-return correlation (computed upstream — see webdash._coin_spot_corr) onto
    the strat grid: two strats on the SAME coin are one bet by construction (struct = 1.0,
    the conservative default for risk); two strats on different coins inherit those coins'
    spot rho. It is an HONEST PROXY — labelled as spot co-movement, an upper bound on the
    co-loss risk — not a substitute for the realized phi once that fills.
  • Redundancy + the effective number of independent groups are driven by LOSS phi (realized)
    AND, separately, by the STRUCTURAL spot rho (immediate). Two runners that crater together
    — or load on the same / a tightly-correlated underlying — are the duplicate risk you must
    not double-arm.

Pure, stdlib-only (no numpy) — paper stays dependency-free. The spot correlation is passed
IN (coin_of / coin_corr); this module never touches the network.
"""

from __future__ import annotations

import math

from .journal import window_pnl

# A correlation needs enough paired windows to mean anything; below this we report the
# overlap count but not a coefficient (honest small-n, like the win-rate tripwire).
MIN_OVERLAP = 10
# Loss phi additionally needs enough LOSSES on each side — favorites win ~90%, so a phi
# computed off 1-2 losses is pure noise. Thin even at 5; the co-loss counts are shown raw.
MIN_LOSSES = 5
# At/above this |phi| two runners lose together enough to be one bet — don't arm both.
REDUNDANT_RHO = 0.7


def pearson(xs: list, ys: list) -> float | None:
    """Pearson correlation of two equal-length series, or None when undefined: too few
    points, or one series is constant (no variance to correlate). Used for both the P&L
    co-movement and — on 0/1 loss indicators — the simultaneous-loss phi coefficient."""
    n = len(xs)
    if n < MIN_OVERLAP:
        return None
    mx, my = sum(xs) / n, sum(ys) / n
    sx = math.sqrt(sum((x - mx) ** 2 for x in xs))
    sy = math.sqrt(sum((y - my) ** 2 for y in ys))
    if sx == 0 or sy == 0:
        return None
    cov = sum((x - mx) * (y - my) for x, y in zip(xs, ys))
    return cov / (sx * sy)


def loss_stats(xs: list, ys: list) -> dict:
    """Simultaneous-loss stats for two aligned per-window P&L series (a window is a loss
    iff its net P&L < 0). Returns the phi correlation of the loss indicators plus the raw
    counts: losses each side (la/lb), windows BOTH lost (both), and the lift of co-losses
    over what independence would predict (>1 = losses cluster = dangerous for the book)."""
    la_ind = [1 if x < 0 else 0 for x in xs]
    lb_ind = [1 if y < 0 else 0 for y in ys]
    la, lb = sum(la_ind), sum(lb_ind)
    both = sum(1 for p, q in zip(la_ind, lb_ind) if p and q)
    n = len(xs)
    phi = (pearson(la_ind, lb_ind)
           if (n >= MIN_OVERLAP and la >= MIN_LOSSES and lb >= MIN_LOSSES) else None)
    exp = (la / n) * (lb / n) * n if n else 0.0   # co-losses expected if independent
    lift = (both / exp) if exp > 0 else None
    return {"loss": phi, "la": la, "lb": lb, "both": both, "lift": lift}


def _independent_groups(names: list, pairs: list) -> int:
    """Effective number of independent bets = connected components once every redundant
    pair (|phi| >= REDUNDANT_RHO, i.e. they LOSE together) is merged. Union-find; a runner
    with no redundant partner stays a singleton (the conservative, honest default — no
    measured co-loss means treat it as its own bet)."""
    parent = {n: n for n in names}

    def find(x):
        while parent[x] != x:
            parent[x] = parent[parent[x]]
            x = parent[x]
        return x

    for a, b in pairs:
        parent[find(a)] = find(b)
    return len({find(n) for n in names})


def _struct_rho(a, b, coin_of, coin_corr):
    """Structural spot co-movement between two strats: 1.0 if they bet the SAME coin (one
    bet by construction — the conservative risk default), else their coins' spot-return
    Pearson (or None if not measurable / no coin data). Symmetric; diagonal handled by the
    caller. coin_of = {name: coin}; coin_corr = {coin: {coin: rho}} (see webdash)."""
    if not coin_of:
        return None
    ca, cb = coin_of.get(a), coin_of.get(b)
    if not ca or not cb:
        return None
    if ca == cb:
        return 1.0
    return (coin_corr or {}).get(ca, {}).get(cb)


def correlation_matrix(runners: dict, coin_of: dict | None = None,
                       coin_corr: dict | None = None) -> dict:
    """Build the live correlation payload from {name: trades}.

    coin_of / coin_corr (optional) enable the STRUCTURAL spot lens: coin_of maps each
    runner to its underlying (e.g. 'zlead-btc-15m' -> 'btc'); coin_corr is the precomputed
    coin×coin spot-return correlation {coin: {coin: rho}} (webdash._coin_spot_corr). Absent
    (the default, e.g. unit tests), every struct field is None and the payload is unchanged.

    Returns:
      names    runner names, in input order (rows == cols)
      windows  per-runner count of resolved windows (the diagonal's n)
      losses   per-runner count of LOSING windows (the simultaneous-loss denominator)
      matrix   names×names of {rho, loss, struct, n, la, lb, both, lift}: rho = full-P&L
               Pearson (secondary), loss = simultaneous-loss phi (primary), struct = spot
               co-movement (structural, immediate), n = overlap, la/lb = losses each side,
               both = windows both lost. rho/loss/struct None when undefined.
      redundant  [{a, b, loss, both, la, lb, n}] pairs with |phi| >= REDUNDANT_RHO,
                 |phi| desc — the "lose together, don't double-arm" list.
      n_strats, n_groups   lineup size + effective independent groups (LOSS-driven).
      avg_abs_loss, n_pairs_loss   simultaneous-loss summary (primary).
      avg_abs_rho, n_pairs         full-P&L summary (secondary toggle).
      struct_groups        effective risk factors after merging same-coin / spot-correlated
                           strats (STRUCTURAL-driven; None when no coin data). The sobering
                           count: how few independent bets the lineup REALLY is, available now.
      avg_abs_struct, n_pairs_struct   structural summary over CROSS-coin pairs (same-coin
                           1.0s excluded — this is the spot beta BETWEEN distinct assets).
      min_overlap, min_losses, redundant_rho   thresholds, so the UI can explain itself.
    """
    series = {name: window_pnl(trades) for name, trades in runners.items()}
    names = list(series)
    losses = {n: sum(1 for v in series[n].values() if v < 0) for n in names}

    matrix = []
    redundant = []
    red_pairs = []
    struct_pairs = []
    abs_loss_sum = abs_rho_sum = abs_struct_sum = 0.0
    n_pairs_loss = n_pairs = n_pairs_struct = 0
    for i, a in enumerate(names):
        row = []
        for j, b in enumerate(names):
            if i == j:
                row.append({"rho": 1.0, "loss": 1.0, "struct": 1.0, "n": len(series[a]),
                            "la": losses[a], "lb": losses[a], "both": losses[a], "lift": None})
                continue
            common = sorted(set(series[a]) & set(series[b]))
            xs = [series[a][t] for t in common]
            ys = [series[b][t] for t in common]
            rho = pearson(xs, ys)
            ls = loss_stats(xs, ys)
            struct = _struct_rho(a, b, coin_of, coin_corr)
            row.append({"rho": rho, "struct": struct, "n": len(common), **ls})
            if j > i:                              # each unordered pair once (upper triangle)
                if rho is not None:
                    abs_rho_sum += abs(rho)
                    n_pairs += 1
                if ls["loss"] is not None:
                    abs_loss_sum += abs(ls["loss"])
                    n_pairs_loss += 1
                    if abs(ls["loss"]) >= REDUNDANT_RHO:
                        redundant.append({"a": a, "b": b, "loss": ls["loss"], "both": ls["both"],
                                          "la": ls["la"], "lb": ls["lb"], "n": len(common)})
                        red_pairs.append((a, b))
                if struct is not None:
                    same_coin = coin_of and coin_of.get(a) == coin_of.get(b)
                    if not same_coin:              # the spot beta BETWEEN distinct assets
                        abs_struct_sum += abs(struct)
                        n_pairs_struct += 1
                    if abs(struct) >= REDUNDANT_RHO:   # same-coin (1.0) or tightly co-moving
                        struct_pairs.append((a, b))
        matrix.append(row)

    redundant.sort(key=lambda r: -abs(r["loss"]))
    has_struct = coin_of is not None and coin_corr is not None
    return {
        "names": names,
        "windows": {n: len(series[n]) for n in names},
        "losses": losses,
        "matrix": matrix,
        "redundant": redundant,
        "n_strats": len(names),
        "n_groups": _independent_groups(names, red_pairs),
        "avg_abs_loss": (abs_loss_sum / n_pairs_loss) if n_pairs_loss else None,
        "n_pairs_loss": n_pairs_loss,
        "avg_abs_rho": (abs_rho_sum / n_pairs) if n_pairs else None,
        "n_pairs": n_pairs,
        "struct_groups": _independent_groups(names, struct_pairs) if has_struct else None,
        "avg_abs_struct": (abs_struct_sum / n_pairs_struct) if n_pairs_struct else None,
        "n_pairs_struct": n_pairs_struct,
        "min_overlap": MIN_OVERLAP,
        "min_losses": MIN_LOSSES,
        "redundant_rho": REDUNDANT_RHO,
    }
